module/signature/signer_store.go

```diff
@@ -0,0 +1,67 @@
+package signature
+
+import (
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+// EpochAwareSignerStore implements the SignerStore interface. It is epoch
+// aware, and provides the appropriate threshold signers on a per-view basis,
+// using the database to retrieve the relevant DKG keys.
+type EpochAwareSignerStore struct {
+	epochLookup module.EpochLookup                // used to fetch epoch counter by view
+	keys        storage.DKGKeys                   // used to fetch DKG private key by epoch
+	signers     map[uint64]module.ThresholdSigner // cache of signers by epoch
+}
+
+// NewEpochAwareSignerStore instantiates a new EpochAwareSignerStore
+func NewEpochAwareSignerStore(epochLookup module.EpochLookup, keys storage.DKGKeys) *EpochAwareSignerStore {
+	return &EpochAwareSignerStore{
+		epochLookup: epochLookup,
+		keys:        keys,
+		signers:     make(map[uint64]module.ThresholdSigner),
+	}
+}
+
+// GetThresholdSigner returns the threshold-signer for signing objects at a
+// given view. The view determines the epoch, which determines the DKG private
+// key underlying the signer.
+func (s *EpochAwareSignerStore) GetThresholdSigner(view uint64) (module.ThresholdSigner, error) {
+	epoch, err := s.epochLookup.EpochForView(view)
+	if err != nil {
+		return nil, err
+	}
+	signer, ok := s.signers[epoch]
+	if ok {
+		return signer, nil
+	}
+	privDKGData, err := s.keys.RetrieveMyDKGPrivateInfo(epoch)
+	if err != nil {
+		return nil, err
+	}
+	signer = NewThresholdProvider(encoding.RandomBeaconTag, privDKGData.RandomBeaconPrivKey)
+	s.signers[epoch] = signer
+	return signer, nil
+}
+
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+// SingleSignerStore implements the ThresholdSignerStore interface. It only
+// keeps one signer and is not epoch-aware. It is used only for the
+// bootstrapping process.
+type SingleSignerStore struct {
+	signer module.ThresholdSigner
+}
+
+// NewSingleSignerStore instantiates a new SingleSignerStore.
+func NewSingleSignerStore(signer module.ThresholdSigner) *SingleSignerStore {
+	return &SingleSignerStore{
+		signer: signer,
+	}
+}
+
+// GetThresholdSigner returns the signer.
+func (s *SingleSignerStore) GetThresholdSigner(view uint64) (module.ThresholdSigner, error) {
```

Comment:
I'm wondering if `view` is enough to index the threshold signer in case a malicious block uses an older view. Shouldn't the we use the block ID too/instead ?

Reply:
The way we determine which signer to use (and which epoch we are in) is based on view. If a block is submitted, malicious or not, with a view that falls within a previous epoch, the only reasonable thing we can really do to verify the signature is use the signer from the corresponding epoch.

That being said, if we are up to date, we should discard a proposal with an old view. If we are behind, we can evaluate the proposal as usual and discard it if it is invalid.

Reply:
What I meant by "older view" is a view in the same epoch, but conflicting with an existent one. So we would have 2 different blocks using the same threshold signer.

Reply:
Every block within an epoch uses the same threshold signer.